### PR TITLE
8278268 - (ch) InputStream returned by Channels.newInputStream should have fast path for FileChannel targets

### PR DESCRIPTION
**Pull request by mkarg**
This sub-issue defines the work to be done to implement JDK-8265891 solely for the particular case of FileChannel.transferFrom(ReadableByteChannel), including special treatment of SelectableByteChannel.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8278268](https://bugs.openjdk.org/browse/JDK-8278268): (ch) InputStream returned by Channels.newInputStream should have fast path for FileChannel targets

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/6711/head:pull/6711` \
`$ git checkout pull/6711`

Update a local copy of the PR: \
`$ git checkout pull/6711` \
`$ git pull https://git.openjdk.org/jdk.git pull/6711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6711`

View PR using the GUI difftool: \
`$ git pr show -t 6711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/6711.diff">https://git.openjdk.org/jdk/pull/6711.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:301]: I've been busy so only getting to this PR now.

Since we are guaranteed that src is configured blocking then the only remaining reason for transferFrom to return 0 is when EOF is reached. Can we drop the temporary buffer and just replace the buffer in the block with:
```
            long n;
            while ((n = dst.transferFrom(src, pos, Long.MAX_VALUE)) > 0) {
                pos += n;
            }
```

Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: I would say "yes, for the channels being part of the JRE", and have adopted your proposal in https://github.com/openjdk/jdk/pull/6711/commits/3c1d8af58e65b6aa5d373c69f7e29e1f10010dc1. The only risk I see (and hence I wrote that originally complex code) is that *in theory* anybody is technically able to implement a **custom** `ReadableByteChannel` (as it is a public interface) which does return zero *intermittently* while the end of stream is not reached. `transferFrom` in that case will also return zero. Hence the loop would terminate *(possibly long) before* EOF is reached.

So do we ignore this hypothetical case?

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: The read method is specified so that "It is guaranteed, however, that if a channel is in blocking mode and there is at least one byte remaining in the buffer then this method will block until at least one byte is read". So in blocking mode it means that read returns >=1 or -1/EOF. You are correct that a blocking of implementation of RBC that returns 0 would be problematic. In the case of transferFrom then it would assume the channel it at EOF. In other usages it may cause a loop as the caller might not expect 0. A completely broken implementation might return random values (-2 !!) so I don't think we can reliably defend against all cases. So I think the updated implementation in 3c1d8af5 is okay.

Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: Ah, right, seems I missed that section about the one-byte-guarantee. Great, so I kindly request approval of this PR!

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: > Ah, right, seems I missed that section about the one-byte-guarantee. Great, so I kindly request approval of this PR!

I think Lance is going to help on this. We need to study the test update closely as this test has been causing a lot of issues in our CI (timeouts, hangs, ...).  I think the main issue with the test right now is JDK-8278369 and it may only be on Windows. I haven't had time myself to study it closely.

Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: The test update only allows to wrap the input file stream with a buffer input stream to skip the special file stream case and walk directly into the general stream case.

**Review by LanceAndersen — CHANGES_REQUESTED** [review]
We will have to separate out the added changes into its own test as we are failing sporadically on some of the windows boxes due to  the execution time.

There is also further cleanup that can be done to existing test.  Attached is an example diff based on the proposed changes in this PR
[8278268-TransferTo-Cleanup.patch.txt](https://github.com/openjdk/jdk/files/7728608/8278268-TransferTo-Cleanup.patch.txt)

**Comment by mkarg** [comment]
Is the cleanup essential for this PR?

**Comment by mkarg** [comment]
> We will have to separate out the added changes into its own test as we are failing sporadically on some of the windows boxes due to the execution time.

@LanceAndersen Do I understand correct, *ontop* of the diff you sent me, you also want me to strip all 2GB tests into a separate test class, correct?

**Comment by LanceAndersen** [comment]
> > We will have to separate out the added changes into its own test as we are failing sporadically on some of the windows boxes due to the execution time.
> 
> @LanceAndersen Do I understand correct, _ontop_ of the diff you sent me, you also want me to strip all 2GB tests into a separate test class, correct?

The updates that you made to add an additional test is resulting in sporadic timeouts on Windows.  

- Please revert TransferTo.Java to omit  the changes you made to it as part of this PR
- -  Yes, we should cleanup TransferTo.java  while we are addressing this test area
- Add a new test file to address your changes to ChannelInputStream found in this PR

**Comment by mkarg** [comment]
@LanceAndersen Thanks for the explanation, I (so I hope) did what you asked me for. The original TransferTo test now does not contain the new tests of *this* PR anymore, but there is a new TransferTo2 which *only* contain the new tests of *this* PR. Both contains your proposed cleanups. On my Windows laptop I never have nor had timeout problems.

**Comment by mkarg** [comment]
@LanceAndersen Requested changes are done since three days. Anything else you want me to change?

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java
Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: @AlanBateman Seems Lance has no time (or is possibly on holidays) so I kindly ask for approval again, as everything he asked me for is fixed in this PR since a week already.

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: > @AlanBateman Seems Lance has no time (or is possibly on holidays) so I kindly ask for approval again, as everything he asked me for is fixed in this PR since a week already.

I've been busy and haven't been following the recent discussions about the test. Lance has been busy too but I think he has been looking at the tests. He mentioned that the new TransferTo2 tests actually runs TransferTo due to an issue with the @run tag. I will try to get time to look at the test soon. Once this PR is in integrated then I think JDK-8278369 should be priority, I think that issue may be Windows specific.

Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: Good catch! Silly me. Fixed in https://github.com/openjdk/jdk/pull/6711/commits/fc7d00bf20861276e11e755c6794999eca95049d.

 JDK-8278369: Tried several times on my Windows 11 laptop, but still cannot reproduce. Maybe Lance's boxes are simply overloaded / too slow?

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: > JDK-8278369: Tried several times on my Windows 11 laptop, but still cannot reproduce. Maybe Lance's boxes are simply overloaded / too slow?

I don't know if anyone has reproduced locally, instead it seems to be random Windows machines in our CI. There are differences between Windows client and server editions that periodically cause issues, I don't know if this is the case here.

**Review by LanceAndersen — CHANGES_REQUESTED** [review]
There is additional clean up that can still be done.  There is duplication of some methods and constants between both tests and this could be streamlined to further reduce duplication

I have also attached a diff of  minor cleanup with the tests as they are additional cleanup to the above work.

As Alan mentioned we are seeing failures in approximately 3% of the runs on various windows boxes.  It appears to be an issue with TransferTo.testStreamContents->TransferTo.checkTransferredContents and selectableChannelOutput() as it blocking.   I have not been able to determine the cause. 

[testcleanup-2.patch.txt](https://github.com/openjdk/jdk/files/7764237/testcleanup-2.patch.txt)

**Comment by mkarg** [comment]
@LanceAndersen I assumed the second test will exist only for the time until we fixed the Windows problems, so I did not invest time to reduce duplication so far. So you really want to keep the separated tests forever and want me to reduce duplication?

**Comment by LanceAndersen** [comment]
> @LanceAndersen I assumed the second test will exist only for the time until we fixed the Windows problems, so I did not invest time to reduce duplication so far. So you really want to keep the separated tests forever and want me to reduce duplication?

I would not make any such assumptions especially for tests that create large files.  So yes please address the duplication.  It is not a huge amount of work to  create a base class that the tests extend (you will see examples elsewhere within the various test directories).

**Comment by mkarg** [comment]
@LanceAndersen Ok, will do that. Stay tuned. :-)

**Comment by mkarg** [comment]
@LanceAndersen I have refactored most of the shared code into a common super class. I deliberately left the 2GB+ tests as (mostly) duplicated code until we sorted out the actual origin of the sporadic failures. If you do not want to wait until then, just tell me, so I will consolidate both 2GB+ test functions into a single one.

**Comment by mkarg** [comment]
> As Alan mentioned we are seeing failures in approximately 3% of the runs on various windows boxes. It appears to be an issue with TransferTo.testStreamContents->TransferTo.checkTransferredContents and selectableChannelOutput() as it blocking. I have not been able to determine the cause.

Looking at the fact that selectableChannelOutput() utilizes Channels.newInputStream() it comes to my mind that [Trisha Gee reported](https://trishagee.com/2021/12/21/simple-network-connections-with-java-a-problem/) that Channels.newInputStream() produced an endless hanging of her apparently correct code, and it was gone once she replaced it by Channels.newReader(). While she reproduced that on a Mac, it might be the same cause even on Windows. Trisha currently is working on a reproducer, but she already detected that *her* endless hanging is definitively a thread waiting for the blocking lock. This sounds similar to what you wrote above. So maybe the problem is not inside of TransferTo, but inside of ChannelInputStream?

**Review by LanceAndersen — COMMENTED** [review]
Thank you for the current updates.  We are starting to move in the right direction.

Alan and I had an offline conversation and concluded the following changes should be made:

- Rename the base class as it is not an Abstract class so the name is confusing
- The base class should only contain constants and helper/utility methods. The actual  tests should reside  in the test classes
- That the tests should be broken out  as
  -  A separate test class for each large file test with a more descriptive class name
  -  A  primary TransferTo test class which includes the remaining tests

After the next set of updates, I will kick off some additional Mach5 runs

I realize that we are spending a lot of time on the tests but they are a key aspect of making sure we have adequate coverage for changes such as this and are easy to maintain and understand by future maintainers/developers

**Comment by mkarg** [comment]
@LanceAndersen I will do as you requested, but due to the holidays it will need a few days. Stay tuned and Merry Christmas!

**Comment by AlanBateman** [comment]
> Looking at the fact that selectableChannelOutput() utilizes Channels.newInputStream() it comes to my mind that [Trisha Gee reported](https://trishagee.com/2021/12/21/simple-network-connections-with-java-a-problem/) that Channels.newInputStream() produced an endless hanging of her apparently correct code, and it was gone once she replaced it by Channels.newReader(). While she reproduced that on a Mac, it might be the same cause even on Windows. Trisha currently is working on a reproducer, but she already detected that _her_ endless hanging is definitively a thread waiting for the blocking lock. This sounds similar to what you wrote above. So maybe the problem is not inside of TransferTo, but inside of ChannelInputStream?

I think the hang in JDK-8278369 is specific to the Unix domain socket implementation of Pipe on Windows 10+ and Windows Server 2019+. More specifically, I think the issue is closing the sink after a large number of bytes has been written does not always cause a reader on the source side to wakeup with EOF. As a test, I changed the implementation to use TCP sockets unconditional (so it works like Windows 8 or Windows Server 2012 or 2016) and I cannot duplicate the issue. We need to look closer at this in the new year.

I looked briefly at the chat server and it has a number of issues.  There is a ArrayList mutated and read from several threads without synchronization. There are several issues that are possible when a chat client does not keep up, this will eventually cause the chat server to hang (a thread blocked on the PrintWriter) and all other threads will eventually blocking trying to echo the messages. So from a quick look, I don't think it is anything to do with the issues we are discussing here.

**Comment by mkarg** [comment]
@AlanBateman Thanks a lot for investing your time. At least it is good news for TransferTo. ;-)

@trishagee Can you please double-check that your problem actually is solved *solely* by replacing streams by writers?

**Comment by mkarg** [comment]
@LanceAndersen I'm done with the requested changes. For "the remaining tests" I kept the separation into TransferTo (existing tests) and TransferTo2 (new tests of this PR) -- or do you want me to consolidate TransferTo2 into TransferTo now that Alan identified the source of trouble being the Unix Domain Sockets?

**Comment by trishagee** [comment]
@mkarg yes, I can confirm the only change I made was to substitute Channels.newOutputStream with Channels.newWriter, and similarly InputStream/Reader.

The code can be found here: [socket-locking](https://github.com/trishagee/socket-locking/). There's two different server classes, and a test case - the test works as expected with Writer/Reader and blocks with the Streams implementation (MacOS Big Sur, adopt-openjdk-11.0.11/)

**Comment by mkarg** [comment]
> I looked briefly at the chat server and it has a number of issues. There is a ArrayList mutated and read from several threads without synchronization. There are several issues that are possible when a chat client does not keep up, this will eventually cause the chat server to hang (a thread blocked on the PrintWriter) and all other threads will eventually blocking trying to echo the messages. So from a quick look, I don't think it is anything to do with the issues we are discussing here.

@AlanBateman I found [the cause of the problem](https://github.com/trishagee/socket-locking/pull/1#issuecomment-1002187644). The problem is that ChannelInputStream and ChannelOutputStream share the same blocking lock when created for the same IP connection, which is pretty straightforward, as a SocketChannel is bidirectional by nature, but has only one blocking setting for both directions. This shared lock is not documented anywhere, so how should an application programmer know about this? This should *either* be documented in the Channels JavaDocs, or (even better) there should be separate blocking modes per direction. FYI, the channel reader does not produce the deadlock as it does not care about the blocking mode at all but simply expects non-blocking mode so it runs in an active loop (squandering power). Maybe we should align both implementations, reader and stream? WDYT?

**Comment by AlanBateman** [comment]
> @AlanBateman I found [the cause of the problem](https://github.com/trishagee/socket-locking/pull/1#issuecomment-1002187644). The problem is that ChannelInputStream and ChannelOutputStream share the same blocking lock when created for the same IP connection, which is pretty straightforward, as a SocketChannel is bidirectional by nature, but has only one blocking setting for both directions. This shared lock is not documented anywhere, so how should an application programmer know about this? This should _either_ be documented in the Channels JavaDocs, or (even better) there should be separate blocking modes per direction. FYI, the channel reader does not produce the deadlock as it does not care about the blocking mode at all but simply expects non-blocking mode so it runs in an active loop (squandering power). Maybe we should align both implementations, reader and stream? WDYT?

Good sleuthing! This is a design bug that goes back to JSR-51/Java 1.4 (https://bugs.openjdk.java.net/browse/JDK-4509080). SelectableChannel specifies blockingLock, which is okay for adaptors and the like when they are readable or writable, but not both.  The socket adaptors have been re-implemented so as not to require the blockingLock during read/write operations so this limits the issue to the interop support that Channels provides. It's fixable with a bit of surgery. A first attempt would be for Channels.newInputStream to return sc.socket().getInputStream() when the channel is a SocketChannel. This requires dealing with cases where someone attempts to get an input stream from an unconnected or closed channel.  A second attempt would be to add sun.nio.ch.SocketInputStream
```
package sun.nio.ch;

import java.io.IOException;
import java.io.InputStream;

public class SocketInputStream extends InputStream {
    private final SocketChannelImpl sc;
    public SocketInputStream(SocketChannelImpl sc) {
        this.sc = sc;
    }
    @Override
    public int read() throws IOException {
        byte[] a = new byte[1];
        int n = read(a, 0, 1);
        return (n > 0) ? (a[0] & 0xff) : -1;
    }
    @Override
    public int read(byte[] b, int off, int len) throws IOException {
        return sc.blockingRead(b, off, len, 0);
    }
    @Override
    public int available() throws IOException {
        return sc.available();
    }
    @Override
    public void close() throws IOException {
        sc.close();
    }
}
```
The main question is whether it's worth doing.

**Comment by mkarg** [comment]
> Good sleuthing! This is a design bug that goes back to JSR-51/Java 1.4 (https://bugs.openjdk.java.net/browse/JDK-4509080) ... The main question is whether it's worth doing.

Thank you for your ideas. On a gut level I would say I will try it out and when the result smells fine I will open a PR for further discussion. What intrigues me most is that solving this problem also would effectively result in potentially higher throughput of the overall system, as read and write are not mutual exclusive anymore. So from a first glimpse I would say it is worth doing it, but I need to give it more in-depth analysis.

**Comment by mkarg** [comment]
@AlanBateman @LanceAndersen Back to the original thread of this PR, are there more change requests or is it fine for merging?

**Comment by AlanBateman** [comment]
> @AlanBateman @LanceAndersen Back to the original thread of this PR, are there more change requests or is it fine for merging?

Lance was looking at the test update, I'm sure he'll get back to it after the holidays.

**Comment by LanceAndersen** [comment]
> @LanceAndersen I'm done with the requested changes. For "the remaining tests" I kept the separation into TransferTo (existing tests) and TransferTo2 (new tests of this PR) -- or do you want me to consolidate TransferTo2 into TransferTo now that Alan identified the source of trouble being the Unix Domain Sockets?

I haven't gone through this in detail, but we want to merge The remaining TransferTo2 tests into TransferTo.  For the  large File tests, any duplicate code can also be included in the base class.

I will look at this more when I am back from the holiday break

**Comment by AlanBateman** [comment]
@mkarg Do you have any benchmarks/results for the change? I'm asking because FileChannel transferTo is implemented (where possible) on sendfile/equivalent so fast for the file -> socket and file -> file cases. The changes in this PR make use of FileChannel transferFrom so may not be significantly better than the default implementation.

**Comment by mkarg** [comment]
I will set up an explicit benchmark for this case.

**Comment by mkarg** [comment]
Please keep open, still worken in it.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
Please keep this PR open; I will continune work on it soon.

**Comment by mkarg** [comment]
> @mkarg Do you have any benchmarks/results for the change?

@AlanBateman It needed several months, but finally I found the time to provide the requested performance numbers:
```
TransferToPerformance.inputStreamLoop    104857  thrpt   25  20557.862 ± 1539.845  ops/s
TransferToPerformance.transferFrom       104857  thrpt   25  22093.131 ± 1178.483  ops/s

TransferToPerformance.inputStreamLoop   1048576  thrpt   25   2366.391 ±  189.846  ops/s
TransferToPerformance.transferFrom      1048576  thrpt   25   2553.160 ±   43.617  ops/s

TransferToPerformance.inputStreamLoop  10485760  thrpt   25    197.128 ±    8.086  ops/s
TransferToPerformance.transferFrom     10485760  thrpt   25    202.492 ±    6.436  ops/s
```
These numbers undoubtedly proof that `transferFrom` is *always faster* than `InputStream::transferTo`'s original loop, for *all* tested stream lengths.

> @mkarg I'm asking because FileChannel transferTo is implemented (where possible) on sendfile/equivalent so fast for the file -> socket and file -> file cases. The changes in this PR make use of FileChannel transferFrom so may not be significantly better than the default implementation.

@AlanBateman The existing cases only covered file-to-something, while the current PR optimizes the reverse something-to-file way. As the above numbers proof that (why ever) `transferFrom` is *always faster* than the original loop of `InputStream::transferTo`, I will continue work on this PR ASAP.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java
Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: @LanceAndersen AFAIK the problem was not originally caused by this PR itself, but in fact by a problem inside of Windows, and recent JDKs workaround that problem somehow meanwhile. So it should be safe to continue with this PR now. If I am wrong please let me know. Thanks.

**Comment by mkarg** [comment]
> @mkarg Please do not rebase or force-push to an active PR as it invalidates existing review comments.

Justification: I have updated this PR by rebasing the 8278268 branch ontop of the master branch. This was needed because the original code had merge conflicts, since it was originally authored ontop of JDK 18.

**Comment by mkarg** [comment]
@AlanBateman @LanceAndersen Kindly re-requesting review of this PR. Thanks! :-)

**Comment by AlanBateman** [comment]
Which platform was used to generate the performance data? Also what was the source stream? Just asking as FileChannel.transferFrom uses copy_file_range for file-to-file (Linux), memory mapping for file-to-file, and a copy loop for other cases.

In any case, the changes to ChannelInputStream.transferTo does the right thing. Most of the change in the PR is the tests. I don't have time to look at those closely right now but Lance or Brian might have cycles.

**Comment by mkarg** [comment]
> Which platform was used to generate the performance data?

Ubuntu 20.04.5 LTS (x86_64)

> Also what was the source stream?

`Channels.newInputStream(Channels.newChannel(new ByteArrayInputStream(byte[size])))`

> Just asking as FileChannel.transferFrom uses copy_file_range for file-to-file (Linux), memory mapping for file-to-file, and a copy loop for other cases.

Yes, but apparently `FileChannel.transferFrom` is doing such copy loop more efficient than `InputStream.transferTo`'s loop.

> In any case, the changes to ChannelInputStream.transferTo does the right thing. Most of the change in the PR is the tests. I don't have time to look at those closely right now but Lance or Brian might have cycles.

Thanks. Please remember that the code was already reviewed and test-driven by Lance a year ago and just was held back due to the bug in Windows's Socket implementation, *discovered by* this PR's test.

**Comment by mkarg** [comment]
@LanceAndersen "Ping" :-)

**Comment by mkarg** [comment]
@AlanBateman Kindly requesting to continue this review.

**Comment by mkarg** [comment]
@AlanBateman @LanceAndersen Kindly asking to resume this MR. Thanks! :-)

**Comment by bplb** [comment]
> @AlanBateman @LanceAndersen Kindly asking to resume this MR. Thanks! :-)

It's in my queue to help out with this.

**Comment by bplb** [comment]
@mkarg: Comments from @LanceAndersen on 2 January 2022:

> I haven't gone through this in detail, but we want to merge The remaining TransferTo2 tests into TransferTo. For the large File tests, any duplicate code can also be included in the base class.

Unless I missed it somehow, I don't see that this has been addressed. Is there an intent to perform the suggested test merging and duplicate code consolidation?

File: test/jdk/java/nio/channels/Channels/TransferTo2.java
Comment (bplb) [inline test/jdk/java/nio/channels/Channels/TransferTo2.java:2]: The copyright years need to be updated per the usual convention in all files in this PR.

**Comment by mkarg** [comment]
> @mkarg: Comments from @LanceAndersen on 2 January 2022:
> > I haven't gone through this in detail, but we want to merge The remaining TransferTo2 tests into TransferTo. For the large File tests, any duplicate code can also be included in the base class.
>
> Unless I missed it somehow, I don't see that this has been addressed. Is there an intent to perform the suggested test merging and duplicate code consolidation?

@bplb I would be happy to continue with merging and removing duplicate code, but I have not yet started this, because @LanceAndersen wrote "I haven't gone through this in detail ... I will look at this more when I am back from the holiday break". So we need to agree *to which extend* we merge both tests (or we need to agree that both files shall become really just one file, as it was originally, before Lance asked to split them). Some background: Originally there was just one file, and it was split into two by request of Lance. So I am not in the position to decide on my own, *which* parts to merge, and *which* parts to keep separated. Can you please give me instructions, which parts you want to keep separated (or confirm that the target really is to just have on single file in the end)? Then I will merge the rest and remove duplicates. Thank you! :-)

File: test/jdk/java/nio/channels/Channels/TransferTo2.java
Comment (mkarg) [reply test/jdk/java/nio/channels/Channels/TransferTo2.java:2]: Thanks a lot for this hint! I will do that at thanks. Apparently this happened because the PR was created in 2021 actually.

Comment (mkarg) [reply test/jdk/java/nio/channels/Channels/TransferTo2.java]: I have pushed commit https://github.com/openjdk/jdk/pull/6711/commits/4bd0d545414756276dff6d2d4f43167cc451425f which updates the copyright year of all touched files.

Comment (bplb) [reply test/jdk/java/nio/channels/Channels/TransferTo2.java]: Looks correct now.

**Comment by bplb** [comment]
Before anything else, you might consider merging the current master into your branch. Aside from the copyright year change of the most recent commit, the branch is over two and a half months out of date.

**Comment by bplb** [comment]
In the test classes `TransferTo` and `TransferTo2`, the methods `testNullPointerException` and `testStreamContents` are identical aside from expecting different `DataProvider`s. The contents of the methods could be abstracted out to common methods called by different `Test`s which expect different `DataProvider`s.

The test classes `TransferTo_2GB_transferFrom` and `TransferTo_2GB_transferTo` comprise roughly 120 lines of code each, but a diff of their contents reveals little in the way of substantive differences aside from jtreg tags, comments, and names:
```
96,99c95,97
<                 // performing actual transfer, effectively by multiple invocations of
<                 // FileChannel.transferFrom(ReadableByteChannel)
<                 try (InputStream inputStream = Channels.newInputStream(Channels.newChannel(
<                         new BufferedInputStream(Files.newInputStream(sourceFile))));
---
>                 // perform actual transfer, effectively by multiple invocations
>                 // of FileChannel.transferTo(WritableByteChannel)
>                 try (InputStream inputStream = Channels.newInputStream(FileChannel.open(sourceFile));
```
Perhaps something along these lines might work:
* rename `TransferToBase` to for example to `TransferBase` or `TransferFromToBase`;
* rename `TransferTo2` to `TransferFrom`;
* rename `TransferTo_2GB_transferX` classes to `TransferX_2GB`;
* refactor common code to `TransferBase` along the lines described above;
* refactor child classes as needed to use the common code.

Assuming the foregoing makes sense, there would remain a total of four tests. It might be possible also to merge `TransferFrom` and `TransferTo` into a single test, but I think the two tests `*_2GB*` should remain separate as they likely each have a sufficiently long run time.

**Comment by mkarg** [comment]
> Before anything else, you might consider merging the current master into your branch. Aside from the copyright year change of the most recent commit, the branch is over two and a half months out of date.

I have rebased this PR branch ontop of the current master branch, so it is up-to-date-now.

**Comment by mkarg** [comment]
@bplb Before I stark to refactor anything, it makes sense to briefly discuss your recent proposal:
> In the test classes `TransferTo` and `TransferTo2`, the methods `testNullPointerException` and `testStreamContents` are identical aside from expecting different `DataProvider`s. The contents of the methods could be abstracted out to common methods called by different `Test`s which expect different `DataProvider`s.

This is correct and I have no objections to that, but on the other hand, there is no need to keep the separate `TransferTo2`  class separated, as the reason for the existence of `TransferTo2` was a bug somewhere else in OpenJDK that does not
exist anymore (for reference, the bug was in the Pipe implementation on Windows). Do that fact, Lance proposed to *merge* these files:
> I haven't gone through this in detail, but we want to merge The remaining TransferTo2 tests into TransferTo. 

So can you please briefly **confirm** that I *shall not* merge `TransferTo2` into `TransferTo` as *Lance* proposed (hence `TransferTo2` will not be a *separate* class anymore) but **instead** I *only shall* abstract `testNullPointerException` and `testStreamContents` as *you* proposed (hence keep `TransferTo2` as a *separate* class)? Thanks.

>...I think the two tests `*_2GB*` should remain separate as they likely each have a sufficiently long run time.

I confirm that we had intentionally kept these two tests in separate classes due to the long run time of each, so I agree to keep them *as separate* classes, but I will reduce *duplicate* code.

I need to mention that the **file names** *should not* be changed according to you proposed, as `*_transferFrom` is actually *not* testing the method `TransferFrom`, but it is testing the method `TransferTo`'s implementation which *internally uses* `transferFrom`. Hence it would be wrong w.r.t to the general test file naming convention to rename the test to `TransferFrom`. 

So can you please briefly **confirm** that I am correct with this objection and the 2GB files *should not* be renamed?

**Comment by bplb** [comment]
> So can you please briefly **confirm** that I _shall not_ merge `TransferTo2` into `TransferTo` as _Lance_ proposed (hence `TransferTo2` will not be a _separate_ class anymore) but **instead** I _only shall_ abstract `testNullPointerException` and `testStreamContents` as _you_ proposed (hence keep `TransferTo2` as a _separate_ class)? Thanks.

That is fine for now. It will be possible to revisit merging at a later time if it appears reasonable.

> > ...I think the two tests `*_2GB*` should remain separate as they likely each have a sufficiently long run time.
> 
> I confirm that we had intentionally kept these two tests in separate classes due to the long run time of each, so I agree to keep them _as separate_ classes, but I will reduce _duplicate_ code.

Sounds good.

> So can you please briefly **confirm** that I am correct with this objection and the 2GB files _should not_ be renamed?

That is fine.

**Comment by mkarg** [comment]
@bplb I am done with the requested changes. There is no duplicated code anymore. :-)

File: test/jdk/java/nio/channels/Channels/TransferToBase.java
Comment (bplb) [inline test/jdk/java/nio/channels/Channels/TransferToBase.java:222]: Should this code perhaps allow for a short write, i.e., `FileChannel::write` returning a value less than `rndBytes.length`? For example, something like this:
```
                    long pos = initPos;
                    while (pos < BYTES_WRITTEN) {
                        int len = Math.toIntExact(BYTES_WRITTEN - pos);
                        if (len > BYTES_PER_WRITE)
                            len = BYTES_PER_WRITE;
                        byte[] rndBytes = createRandomBytes(len, 0);
                        ByteBuffer src = ByteBuffer.wrap(rndBytes);
                        pos += fc.write(src);
                    }
```

Comment (mkarg) [reply test/jdk/java/nio/channels/Channels/TransferToBase.java]: I have incorporated your proposed code change.

**Comment by mkarg** [comment]
@bplb Your requested changes are done since one week. Kindly asking to proceed with this PR. Thanks. :-)

**Comment by bplb** [comment]
> @bplb Your requested changes are done since one week. Kindly asking to proceed with this PR. Thanks. :-)

Indeed I was intending to approve this afternoon. I read through it all and the tests look much better now.

**Review by bplb — APPROVED** [review]

**Pull request closed without merging**
